### ISSUE TITLE
LOST DATA git push overwrites my commits

### ISSUE THREAD

**Issue by zrzka** [ywangd/stash#268]
Test scenario (I did it in the [stash-git-test](https://github.com/zrzka/stash-git-test) repository):

1. Create empty repository
2. Clone it in the StaSh
3. Create README.md
4. Commit it with StaSh (First commit)
5. Go to GitHub website, modify README.md, commit & push (Second commit, everything on the GitHub's website)
6. Return to StaSh
7. Modify README.md, commit & push (Third commit)
8. Go to GitHub website and check commits -> Second commit (5. is gone)
9. Pull (or fetch & merge) in StaSh
10. Issue git log
11. Second commit is really gone

I'm loosing commits in this way pretty often when I'm playing with readme for example and forgot to pull on iPad. If would expect this:

```sh
TO BE COPIED ON MBP
```

**Comment by jsbain** [comment]
Are you using ssh: to push?
It used to be that this scenario caused the dreaded Pipe Error, rather than overwriting, but if using ssh it would happily overwrite.  
I think dulwich is (or was) pushing in a way that looks like a rebase -- as far as I understand, there is no difference in the protocol between a rebase and a push-- i suspect in desktop git, the check is done by the client.

**Comment by zrzka** [comment]
Yup, use ssh. I'm away till Sunday, will check what desktop git does exactly when I'll be back.

**Comment by jsbain** [comment]
desktop git warns you, pretty sure.  

The temporary solution is to do a git fetch prior to push, (and in fact call fetch from within stash git push for the specified remote).  Stash git does check if a push is safe (remote's HEAD is an ancestor of local HEAD), but it does that based on the last fetch result.

**Comment by jsbain** [comment]
i made some progress extracting gittle, to allow using up to date dulwich., hopefully have something this weekend.

The eventual solution to the push issue, porcelain has a ls_remote, which allows one to query the remote without actually doing a complete fetch; so, this would allow checking whether the local is up to date (remote sha is in local object store, and whether it is a direct ancestor of the current commit).

**Comment by jsbain** [comment]
btw, in case anything important was lost, you should be able to find lost commits using this method (curl even works in stash)
https://objectpartners.com/2014/02/11/recovering-a-commit-from-githubs-reflog/
